driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -166,29 +161,15 @@ private boolean hasTimedOutForCommandExecution() {
      * @return timeout to use.
      */
     public long timeoutOrAlternative(final long alternativeTimeoutMS) {
-        Long timeoutMS = timeoutSettings.getTimeoutMS();
-        if (timeoutMS == null) {
-            return alternativeTimeoutMS;
-        }
-        return timeoutRemainingMS();
-    }
-
-    /**
-     * Calculates the minimum timeout value between two possible timeouts.
-     *
-     * @param alternativeTimeoutMS the alternative timeout
-     * @return the minimum value to use.
-     */
-    public long calculateMin(final long alternativeTimeoutMS) {
-        Long timeoutMS = timeoutSettings.getTimeoutMS();
-        if (timeoutMS == null) {
-            return alternativeTimeoutMS;
-        } else if (timeoutMS == 0) {
+        if (timeout == null) {
```

Comment:
This change assumes that it is impossible for timeout to be null when `timeoutSettings.getTimeoutMS()` returns a non-null value. Please confirm.

Reply:
It appears that in practice, based on usages, timeout is never null in such cases. The constructors are always invoked where the timeout is derived via `startTimeout`, or have a non-null value. Also, if this were not the case, then I think `resetTimeout` has a bug, because it derives the timeout based on `timeoutSettings.getTimeoutMS`.

Refactored a bit.